File: opencv_learn/keypoints.py
"""
keypoints 关键点检测
"""

from typing import Literal

import cv2
from cv2.typing import MatLike

from opencv_learn.cv2_utils import PopstarAhri, img_show

KeyPointsType = Literal["sift", "orb"]


@img_show("keypoints", cv2.WINDOW_FREERATIO)
def keypoints(img: MatLike, keypoints_type: KeyPointsType) -> MatLike:
    gray = cv2.cvtColor(img, cv2.COLOR_BGR2GRAY)
    model = {"sift": cv2.SIFT_create, "orb": cv2.ORB_create}.get(keypoints_type)(20000)
    keypoints = model.detect(gray, None)
    cv2.drawKeypoints(img, keypoints, img, (0, 255, 0))
    return img


def main():
    keypoints(PopstarAhri, "sift")
    keypoints(PopstarAhri, "orb")


if __name__ == '__main__':
    main()

File: opencv_learn/keypoints.cpp
/**
 * @file keypoints.cpp
 * @date 2024/07/17
 * @author Sokyoei
 *
 *
 */

#include <filesystem>
#include <iostream>
#include <vector>

#include "Asuka.hpp"

#include <opencv2/features2d.hpp>
#include <opencv2/opencv.hpp>

#include "Ahri/Asuka/opencv_color.hpp"

namespace Ahri {
enum class KeyPointsType {
    SIFT,
    ORB,
};

void keypoints(cv::Mat& img) {
    cv::Mat gray_img;
    cv::cvtColor(img, gray_img, cv::COLOR_BGR2GRAY);
    auto model = cv::SIFT::create(20000);
    std::vector<cv::KeyPoint> keypoints;
    model->detect(img, keypoints);
    cv::drawKeypoints(img, keypoints, img, GREEN);
    cv::namedWindow("keypoints", cv::WINDOW_FREERATIO);
    cv::imshow("keypoints", img);
    cv::waitKey();
    cv::destroyAllWindows();
}
}  // namespace Ahri

int main(int argc, char const* argv[]) {
    auto img_path = std::filesystem::path(SOKYOEI_DATA_DIR) / "Ahri/Popstar Ahri.jpg";
    auto img = cv::imread(img_path.string());
    Ahri::keypoints(img);
    return 0;
}
